File: solutions/2958-length-of-longest-subarray-with-at-most-k-frequency/sliding-window.js
// time: O(n)
// space: O(n)

/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
const maxSubarrayLength = (nums, k) => {
    const n = nums.length;
    const freq = {};
    let left = 0;
    let right = 0;
    let result = 0;

    for (; right < n; right++) {
        const rightNum = nums[right];

        freq[rightNum] ??= 0;
        freq[rightNum]++;

        while (freq[rightNum] > k) {
            const leftNum = nums[left];

            freq[leftNum]--;
            left++;
        }

        const length = right - left + 1;
        result = Math.max(result, length);
    }

    return result;
};

File: solutions/2958-length-of-longest-subarray-with-at-most-k-frequency/sliding-window.cpp
// time: O(n)
// space: O(n)

#include <algorithm>
#include <unordered_map>
#include <vector>

using namespace std;

class Solution {
  public:
    int maxSubarrayLength(vector<int> nums, int k) {
        const int n = nums.size();
        unordered_map<int, int> m;
        int left = 0;
        int right = 0;
        int result = 0;

        for (; right < n; right++) {
            const int rightNum = nums[right];
            m[rightNum]++;

            while (m[rightNum] > k) {
                const int leftNum = nums[left];
                m[leftNum]--;
                left++;
            }

            result = max(result, right - left + 1);
        }

        return result;
    }
};
